### PR TITLE
Fixed the mess and have the VMM in a PR worthy state

### LINKED ISSUE #11: Create a virtual memory manager for the x86_64 ISA
Create a virtual memory manager for the x86_64 ISA that can do the following:

- [x] ~~Enable the use of PCIDs~~*
- [x] Create a page map
- [x] ~~Implement `Clone` and `Drop` for page map~~*
- [x] Load a page map
- [x] Map normal 4 KiB pages
- [x] Unmap normal pages
- [x] Map large pages
- [x] Unmap large pages
- [x] Detect CPU support for huge pages
- [x] Map huge pages
- [x] Unmap huge pages
- [x] ~~Handle page fault exceptions~~**
- [x] ~~Make use of the page attribute table (PAT) to to set memory attributes for different regions~~***

*Postponed until there are multiple PageMaps
**Postponed until there is a userspace
***Moved to a separate issue due to the possibility of breaking the bootloader provided PageMap

### PR DESCRIPTION
**Pull request by mdpcs3544**
Implemented only the parts of the VMM that would be useful for the kernel at this point in development.

Closes #11

### CODE REVIEW

File: charlotte_core/src/arch/x86_64/memory/page_map/page_table/mod.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        index: usize,
+    ) -> Result<PhysicalAddress, Error> {
+        let page_paddr = self.table[index].unmap()?;
+        match size {
+            PageSize::Standard => PHYSICAL_FRAME_ALLOCATOR.lock().deallocate(page_paddr)?,
+            PageSize::Large => PHYSICAL_FRAME_ALLOCATOR
+                .lock()
+                .deallocate_contiguous(page_paddr, 512)?,
+            PageSize::Huge => PHYSICAL_FRAME_ALLOCATOR
+                .lock()
```
Comment (Wrench56) [inline charlotte_core/src/arch/x86_64/memory/page_map/page_table/mod.rs:97]: This seems reduntant to me. Just use deallocate_contiguous on a standard page as well. Wouldn't really affect the performance and this whole chunk can be simplified. Well, of course unless later on there will be other things specific to some page size.

**Review by Wrench56 — APPROVED** [review]
LGTM!

`deallocate` seems to be redundant. You could just use `deallocate_contiguous`.

File: charlotte_core/src/arch/x86_64/memory/page_map/page_table/mod.rs (hunk 1 above)
Comment (Wrench56) [reply charlotte_core/src/arch/x86_64/memory/page_map/page_table/mod.rs:97]: For some reason, the part where you use `deallocate` was cut out... scroll a bit up, it is in the same `match`.

File: charlotte_core/src/memory/address.rs [hunk 2]
```diff
@@ -1,14 +1,18 @@
 use core::num::NonZeroUsize;
 use core::ops::Add;
 
+use crate::arch::{Api, ArchApi};
+use crate::memory::pmm::DIRECT_MAP;
+
+const PAGE_SIZE: NonZeroUsize = unsafe { NonZeroUsize::new_unchecked(4096) };
+const PAGE_SHIFT: usize = 12;
+const PAGE_MASK: u64 = !0xfffu64;
```
Comment (GranitBebok) [inline charlotte_core/src/memory/address.rs]: This was already mentioned on discord, but these should be arch dependent

**Comment by mdpcs3544** [comment]
I've withdrawn this PR because even basic tests are failing in the VMM due to address validation code malfunctioning and integer overflow causing the Rust runtime to panic.

Clearly I submitted this PR way too prematurely.

**Pull request closed without merging**